File: src/main/java/no/nav/syfo/service/OversikthendelseService.java
package no.nav.syfo.service;

import no.nav.syfo.consumer.pdl.PdlConsumer;
import no.nav.syfo.domain.AktorId;
import no.nav.syfo.consumer.behandlendeenhet.BehandlendeEnhetConsumer;
import no.nav.syfo.domain.Fodselsnummer;
import no.nav.syfo.domain.model.Mote;
import no.nav.syfo.kafka.producer.OversikthendelseProducer;
import no.nav.syfo.kafka.producer.OversikthendelseType;
import no.nav.syfo.kafka.producer.model.KOversikthendelse;
import org.springframework.stereotype.Service;

import static java.time.LocalDateTime.now;

@Service
public class OversikthendelseService {

    private final BehandlendeEnhetConsumer behandlendeEnhetConsumer;
    private final PdlConsumer pdlConsumer;
    private OversikthendelseProducer oversikthendelseProducer;

    public OversikthendelseService(
            BehandlendeEnhetConsumer behandlendeEnhetConsumer,
            PdlConsumer pdlConsumer,
            OversikthendelseProducer oversikthendelseProducer
    ) {
        this.behandlendeEnhetConsumer = behandlendeEnhetConsumer;
        this.pdlConsumer = pdlConsumer;
        this.oversikthendelseProducer = oversikthendelseProducer;
    }

    public void sendOversikthendelse(Mote mote, OversikthendelseType type) {
        Fodselsnummer sykmeldtFnr = pdlConsumer.fodselsnummer(new AktorId(mote.sykmeldt().aktorId));
        String behandlendeEnhet = behandlendeEnhetConsumer.getBehandlendeEnhet(sykmeldtFnr.getValue(), null).getEnhetId();

        KOversikthendelse kOversikthendelse = KOversikthendelse.builder()
                .fnr(sykmeldtFnr.getValue())
                .hendelseId(type.name())
                .enhetId(behandlendeEnhet)
                .tidspunkt(now()
                ).build();

        oversikthendelseProducer.sendOversikthendelse(mote.uuid, kOversikthendelse);
    }
}

File: src/main/kotlin/no/nav/syfo/oversikthendelse/OversikthendelseService.kt
package no.nav.syfo.oversikthendelse

import no.nav.syfo.consumer.behandlendeenhet.BehandlendeEnhetConsumer
import no.nav.syfo.consumer.pdl.PdlConsumer
import no.nav.syfo.domain.AktorId
import no.nav.syfo.domain.Fodselsnummer
import no.nav.syfo.domain.model.Mote
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class OversikthendelseService(
    private val behandlendeEnhetConsumer: BehandlendeEnhetConsumer,
    private val pdlConsumer: PdlConsumer,
    private val oversikthendelseProducer: OversikthendelseProducer
) {
    fun sendOversikthendelse(mote: Mote, type: OversikthendelseType) {
        val fnr = pdlConsumer.fodselsnummer(AktorId(mote.sykmeldt().aktorId))
        val behandlendeEnhet = behandlendeEnhetConsumer.getBehandlendeEnhet(fnr.value, null).enhetId

        val kOversikthendelse = map2KOversikthendelse(fnr, behandlendeEnhet, type)
        oversikthendelseProducer.sendOversikthendelse(mote.uuid, kOversikthendelse)
    }

    private fun map2KOversikthendelse(
        fnr: Fodselsnummer,
        tildeltEnhet: String,
        oversikthendelseType: OversikthendelseType
    ) = KOversikthendelse(
        fnr = fnr.value,
        hendelseId = oversikthendelseType.name,
        enhetId = tildeltEnhet,
        tidspunkt = LocalDateTime.now()
    )
}
